### ISSUE TITLE
Complete Message Thread From Inbox UI

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8705]
As a Court user, so that I can quickly complete out message threads that don't require any further comment, I want to be able to complete a thread directly from my Inbox UI.

Currently, in order to complete a message thread in DAWSON, a user must:
1) Click into the message
2) Click Complete
3) Click the Complete button on the pop up UI
4) Navigate back to their inbox to continue working their messages

Many times it is evident from the message display in the UI that the thread simply needs to be completed with no additional comment. In order to facilitate quick disposal of these messages, we would like the ability to simply invoke the Complete functionality for a message directly from the Inbox UI where the message itself is displayed. This functionality does not need to include the ability to add a final comment, but needs only perform the completion of the thread exactly as if the user had invoked it the "traditional" way.

To begin with, we would like the functionality to be available to all court users from their My Messages - Inbox view (NOT the Section inbox at this time). 

## Acceptance Criteria
* User can complete a message thread directly from the Inbox view on the My Messages screen
* After completion of a thread in this manner, the user Inbox view is refreshed and retains any prior sorting/filtering applied.

## Story Definition of Ready (updated on 12/23/22)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ] Is framed in business/user need, the value has been addressed.
- [ ] Includes acceptance criteria
- [ ] Has been refined
- [ ] Pre conditions have been satisfied. 

**Process:** 
Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 5-19-22)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [ ] Automated test scripts have been written, including visual tests for newly added PDFs.
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning.
 - [ ] Verify that language for docket record for internal users and external users is identical.
 - [ ] New screens have been added to pa11y scripts.
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html.
 - [ ] READMEs, other appropriate docs, and swagger/APIs fully updated.
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required).
 - [ ] Interactors should validate entities before calling persistence methods.
 - [ ] Code refactored for clarity and to remove any known technical debt.
 - [ ] If new docket entries have been added as seed data to `efcms-local.json`, 3 local s3 files corresponding to that docketEntryId have been added to `web-api/storage/fixtures/s3/noop-documents-local-us-east-1`
 - [ ] Acceptance criteria for the story has been met.
 - [ ] If there are special instructions in order to deploy into the next environment, add them as a comment in the story.
 - [ ] If the work completed for the story requires a reindex without a migration, or any other special deploy steps, apply these changes to the following flexion branches:
   - [ ] experimental1
   - [ ] experimental2
   - [ ] experimental3
   - [ ] experimental4
   - [ ] experimental5
   - [ ] experimental6
   - [ ] develop
 - [ ] Reviewed by UX on a deployed environment.
 - [ ] Reviewed by PO on a deployed environment. Can be deployed to the Court's `test` environment if prod-like data is required. Otherwise deployed to any `experimental` environment.
 - [ ] Deployed to the Court's `staging` environment.

**Comment by katiecissell** [comment]
Pre-refinement questions:
- Should we also be adding Forward, reply, create order to their inbox as well? (Not in near future)
- Also ideally completing a message in one tab should remove it from other tabs

**Comment by swongCO** [comment]
**UX Notes**
Two options to consider. Both would satisfy the AC but one would offer more functionality.
https://www.figma.com/design/tO9qBTpHSGQzaodPfhLVIF/USTC-%7C-10372-Complete-message-from-inbox?node-id=0-1

Option 1:
- Display icon link at the end of the row (use same link in the case message flow)
- Display same alert banner at the top after user clicks the link
- Message is moved to Completed and is not displayed in the Inbox

<img width="1265" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeCt4QkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--ee79dab4f6b1222043bd1439a3e64d86b822e686/Frame%201.png" alt="Frame 1.png" />

<img width="817" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeUd4QkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--626ccf6141f187d15b858e684d2bd66e2873f9e3/Screenshot%202024-05-01%20at%2011.08.46%E2%80%AFAM.png" alt="Screenshot 2024-05-01 at 11.08.46 AM.png" />

___
Option 2:
- Add checkboxes to the table
- Add Complete link above the table (disabled by default)
- If one or more boxes are checked, enable the Complete link
- Use the same checkbox states for header box from docket record download
- Display a pluralized alert banner (or modify the existing one to include pluralization)
- Message(s) is moved to Completed and is not displayed in the Inbox

<img width="1265" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBM1hRQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--14832d85f2e94a167901b3b5dda8f8b27997e559/Option%202.png" alt="Option 2.png" />

**Comment by ttlenard** [comment]
## Test Cases

~~**NOTE: These test cases are for option 1 above.  If developers determine option 2 is the better solution, the test cases below will be updated.**~~

**I've updated these test cases to match option 2 on 7/16/2024 - Tenille**


#### 1) Court user accesses their My Messages inbox; There is now a link to the right side, above the message table that displays "Complete", checkboxes to the left of each message, and a checkbox in the header of the message table.

- Log into DAWSON as a docket clerk. Be sure that the Docket clerk has messages that have been sent to them.
- Navigate to your My Messages Inbox
- Review the page

Expected Results:
- [x] **NEW FUNCTIONALITY** - there is now a "Complete" link located on the right side of the Messages table.
- [x] **NEW FUNCTIONALITY** - there are now checkboxes to the left of each message



#### 2) Court user clicks on a checkbox in the message table; The complete link at the top of the table becomes activated. 

- Continuing from the previous test
- Click on the one of the checkboxes in one of the messages

Expected Results:
- [x] **NEW FUNCTIONALITY** - The complete link at the top of the message table becomes active.


#### 3) Court user clicks on the complete link; message is completed and confirmation banner is displayed.

- Continuing from the previous test
- Click on the Complete link

Expected Results:

- [x] **NEW FUNCTIONALITY** - The page refreshes, there is the green confirmation banner at the top of the screen that indicates that you completed the message.
- [x] **NEW FUNCTIONALITY** - The message that you just completed is no longer displayed in the inbox



#### 4) Court user selects multiple checkboxes in the message inbox (not all), clicks on the complete link; messages selected are completed and confirmation banner is displayed.

- Continuing from the previous test
- Check the box for multiple messages (not all)
- Click on the Complete link

Expected Results:

- [x] **NEW FUNCTIONALITY** - The page refreshes, there is the green confirmation banner at the top of the screen that indicates that you completed the messages.
- [x] **NEW FUNCTIONALITY** - The messages that you just completed is no longer displayed in the inbox


#### 5) Court user checks the "Select all" checkbox in the message header row; All of the checkboxes in the message inbox are checked

- Continuing from the previous test
- Click on the checkbox in the message header row

Expected Results:
- [x] All of the checkboxes for each message are selected.



#### 6) Court user clicks on the Complete link at the top of the table after selecting all; all messages are marked as completed.

- Continuing from the previous test
- Click on the Complete link in the top right of the message table

Expected Results:
- [x] **NEW FUNCTIONALITY** - The page refreshes, there is the green confirmation banner at the top of the screen that indicates that you completed the messages.
- [x] **NEW FUNCTIONALITY** - The messages that you just completed is no longer displayed in the inbox


#### 7) Court user navigates to their completed messages box; The messages that they just completed are at the top of the table.

- Continuing from the previous test
- Click on the Completed message box

Expected Results:
- [x] **NEW FUNCTIONALITY** - The message that you just completed is now in the Completed message box
- [x] **NEW FUNCTIONALITY** - There is no complete link next to any of the messages in the completed message box.
- [x] Completed message displays as per normal.


#### 8) Court user accesses the message that was completed via the new complete link on their message inbox, data about the completed message is displayed in the case messages as per usual.

- Continuing from the previous test
- Navigate to the docket number of the message that you just completed
- Click on the case messages tab
- Click on completed
- Review the completed messages

Expected Results
- [x] The message that you just completed is displayed as per if you would have completed it within the case messages
- [x] After clicking on the message link, the green banner displays when the message was completed per usual
- [x] You can expand the thread as per usual to see the message history



#### 9) Court user adds some filtering and sorting in their Message inbox; Completes a message, Page refreshes with confirmation alert at the top of the page; Sorting/Filtering remains the same.

- Log in as a docket clerk
- Navigate to your My Messages inbox
- Be sure that you have a handful of messages 
- Filter the messages and apply some sorting
- Complete a message

Expected Results
- [x] **NEW FUNCTIONALITY** - Confirmation message that the message was completed is displayed at the top of the page 
- [x] **NEW FUNCTIONALITY** - The Sorting/Filtering that had been applied remains the same (the user does not have to reset their filters/sorting)

 
~~#### 10) Court user that has A LOT of messages scrolls down to the bottom of the page and completes a message, Page refreshes, confirmation banner displays, and the page is displaying the top of the page.~~

~~- Log in as a docket clerk that has a lot of messages (so you have to scroll down the page to see them all)~~
~~- Navigate to the My Messages inbox~~
~~- Scroll down to near the bottom of the page~~
~~- Click Complete~~

~~Expected Results:~~
~~- [ ] **NEW FUNCTIONALITY** - The page refreshes and displays the yellow alert banner at the top of the page~~
~~- [ ] **NEW FUNCTIONALITY** - User can scroll down the page and the message is no longer in the inbox~~




#### 11) Court user navigates to other message boxes, the only message box that has the complete link is the My Messages inbox.

- Continuing from the previous test
- Click on all of the message boxes from the My Messages and Section messages

Expected Results:
- [x] The complete link is ONLY displayed in the users My Messages inbox, no other message box (my messages or Section messages) has the complete link.



#### 12) Regression test: Court user opens a message, marks the message as complete from within the message, message completes as per usual.

- Continuing from the previous test
- Navigate to a message from the My messages inbox
- Do not complete the message from this view, instead, click on the message link to open the message within the case 
- Verify that you can still complete a message as per previous functionality 

Expected Results:
- [x] You can click complete from the Case messages page
- [x] After clicking the complete link above the message, you have the option to add a comment in the pop-up modal as per normal (this is optional)
- [x] You can click complete/cancel in the pop-up modal as per normal


#### 13) Regression test - Completed Message box behaves a per normal with the display of messages completed via the link in the inbox view.

- [ ] Messages completed via the new link are removed from the completed message box as per usual (7 days)


#### 14) Regression test - Court users can view their message screen on a mobile device

- [x] Court users can view their messages while logged in on a mobile device (it doesn't have to be pretty, it just needs to work).


*Repeat tests 1-14 with all court users (besides general)
- [x] Floater
- [x] Petitions Clerk
- [x] Admissions Clerk
- [x] Judge
- [x] Chambers
- [x] Trial Clerk
- [ ] Reporters office
- [x] Docket Clerk
- [x] ADC
- [x] Case Services Supervisor
- [x] Clerk of the Court

**Comment by jasonfu9966** [comment]
- UX prefers option 2, but we will explore both
- stay on the inbox page when message is complete

nice to haves:
- yellow alert banner changed to green success
- animate table row collapsing

**Comment by jasonfu9966** [comment]
a thing to test:
It's possible for multiple messages from the same thread to be in the inbox. what happens if you complete the non-most-recent message in a thread? does it complete the entire thread?
if yes, then the batch completion behavior should be the same. completing any message from a thread should move all the corresponding messages from the inbox to completed.

follow-up UX question: should we let the user know when they're selecting a message that's part of a thread?

**Comment by jasonfu9966** [comment]
spoke with Tenille. it is *not* possible for multiple messages from the same thread to be in the inbox displays threads only; when you open one, you can see all the messages in the thread.

**Comment by swongCO** [comment]
Spoke with devs:
- Will put the success banner (it's green yay) above the filters within the inbox tab since it's related to the page. 
- There will only be one banner at any given time so new ones will replace the previous one in the same session.
- If the user switches tabs or pages, the banner will not display when the user returns to the inbox so it doesn't need to be dismissible.
- Spacing between the table and filters/counter was reduced to be consistent.

**Comment by swongCO** [comment]
Test feedback
BUG: the select all checkbox isn't working properly. Steps to reproduce:

1. in messages, make sure you have more than one
2. check the select all box
3. deselect one of the rows
4. check the select all box again
5. it’s not selecting all

Screen recording

[Private Zenhub Video](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeHJTQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--7801faf812fbe72ca25924930771facf2d25cd1c/Screen%20Recording%202024-07-11%20at%2010.28.21%E2%80%AFAM.mov)

**Comment by swongCO** [comment]
Test feedback
Different bug with the select all checkbox. Steps to reproduce:

1. Log in as Judge Buch (he has plenty of messages)
2. Complete one message
3. Select all using the box
4. Click the box again to deselect all

The box is not deselecting any rows and remains in an empty state.

**Comment by ttlenard** [comment]
Testing feedback:

Looks like we have an issue with Test case 9 above.

1. Apply a filter (can be any filter), in this example, I had 7 messages in my inbox, and I filtered by case status "General Docket - Not at Issue", and my messages displayed 3 messages in my inbox with the filter applied:
<img width="1358" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOFhvQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--998c5841fd96e84478d8c2c0dee3392989d6f77b/image.png" alt="image.png" />
2. Select one checkbox and then click complete:
<img width="1305" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOGJvQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--c765ad0e1aa8753522de5e854e20f33e863dc55d/image.png" alt="image.png" />
3.  Notice that the total messages in your inbox shifts to whatever you have filtered. It forgets about all of the other messages.
<img width="1302" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOHJvQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--2619425f60a76c8a3eb165ca1b95d82e7e9d6aed/image.png" alt="image.png" />
4.  The only way for it to refresh is to click on a different message box, or go somewhere else in the application so that the message inbox refreshes: In this example, I just clicked on the sent box for it to show the proper message number.
<img width="1359" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOHZvQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f86aed1b62e2a3d14eafab49088fe36cdbc815d1/image.png" alt="image.png" />
